Support externalTrafficPolicy=Local for BGP CPlane service VIP advertisement

YutaroHayakawa opened the pull request:
Support `externalTrafficPolicy=local` for BGP CPlane service VIP advertisement. Please see the commit message for more details.

You can test this PR in your environment with below lab. Thanks to the recently added `Chart CI Push` workflow, you don't have to build images.

[lab.tar.gz](https://github.com/cilium/cilium/files/11494605/lab.tar.gz)

```mermaid
flowchart TB
  Router0["Router0 (FRR)"]
  Router0---kind-control-plane["kind-control-plane (Cilium)"]
  Router0---kind-worker["kind-worker (Cilium)"]
```

```
$ tar xvf lab.tar.gz
$ cd lab
$ VERSION=1.14.0-dev-dev.55-HEAD-f00eeee626 make deploy

# Watch what's happening
$ watch -n 1 "docker exec -it clab-bgp-cplane-router0 vtysh -c 'show bgp all wide'"

# Make endpoints on one node empty. This should withdraw the route from kind-worker.
$ kubectl drain kind-worker --ignore-daemonsets

# Restore endpoints. This should restore the route as well.
$ kubectl uncordon kind-worker
```

```release-note
Support externalTrafficPolicy=local for BGP CPlane service VIP advertisement
```

YutaroHayakawa committed:
bgpv1: Support externalTrafficPolicy=local for service advertisement

The semantics of the externalTrafficPolicy=local for service
advertisement. When at least one active endpoint present, we advertise
the LB VIP, otherwise, stop advertisement.

We can track active endpoints by tracking the Endpoints or EndpointSlice
object. They contain the list of the endpoint IPs and nodes. We leverage
a new Resource[*Endpoints] for subscribing the changes. For an efficient
processing, we wrap it with DiffStore and do reconciliation against the
service affected by the endpoint changes.

On every reconciliation, we populate the structure called localServices
from DiffStore store which represents the information that services have
at least one available local endpoint or not.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

rastislavs commented on pkg/bgpv1/manager/reconcile.go:
```diff
@@ -420,29 +426,61 @@ func (r *LBServiceReconciler) lbServiceReconciler(
+	sla := make(svcLocalAvailability)
+endpointsLoop:
+	for _, eps := range r.epDiffStore.List() {
+		svcID := eps.ServiceID
```
Would it make sense to skip this processing for services that do not have `externalTrafficPolicy=Local` here to save some processing, or is it not worth retrieving the service details at this place?

YutaroHayakawa replied:
Yep, I think that makes sense. It should be a single map lookup, so worth to pay that cost I think.

YutaroHayakawa replied:
Done

harsimran-pabla commented on pkg/bgpv1/manager/reconcile.go:
```diff
@@ -420,29 +434,77 @@ func (r *LBServiceReconciler) lbServiceReconciler(
 	}
 
+	// Populate local availability used for externalTrafficPolicy=local handling
+	sla := make(svcLocalAvailability)
```
I'd suggest populating sla should be moved to a new method.

harsimran-pabla commented on pkg/bgpv1/manager/reconcile.go:
```diff
@@ -420,29 +434,77 @@ func (r *LBServiceReconciler) lbServiceReconciler(
+		// we use EndpointSlice. We shouldn't make available to unavailable.
+		if _, found := sla[svcID]; !found {
+			sla[svcID] = false
+		}
```
`sla` map going to contain all services in the cluster which have ServiceExternalTrafficPolicyTypeLocal.

I wonder can we make map to contain services which have ServiceExternalTrafficPolicyTypeLocal and present locally. i.e. making `sla` map as a set of local services only. This reduces overall size of the map and also potentially simplifies the implementation.  

We can lookup in the map if svcID is present before iterating over eps.Backends, if present do not need to go over the backends. 

Pls correct me if you think my understanding is wrong.

YutaroHayakawa replied:
Yep, it sounds like a good idea. I fixed it.

YutaroHayakawa committed:
bgpv1: Deduplicate resources in TestLBServiceReconciler

TestLBServiceReconciler is getting longer and now it's hard to follow
what's the diff between resources used in tests. Extract resource
definition out of the test case definition and deduplicate the fields as
much as possible.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bgpv1: Add test cases for externalTrafficPolicy=Local

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

dylandreimerink reviewed (approved):
:rocket: LGTM!
